### PR TITLE
Closes #6797 Remove !! operator usages from AbstractFetchDownloadService

### LINKED ISSUE #6797: Crash KotlinNullPointerExceptionmozilla.components.feature.downloads.AbstractFetchDownloadService
Related Fenix issue https://github.com/mozilla-mobile/fenix/issues/7303

```
kotlin.KotlinNullPointerException: null
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.useFileStreamScopedStorage(AbstractFetchDownloadService.kt:18)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.useFileStream$feature_downloads_release(AbstractFetchDownloadService.kt:8)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService$performDownload$1.invoke(AbstractFetchDownloadService.kt:20)
    at mozilla.components.concept.fetch.Response$Body.useStream(Response.kt:1)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.performDownload$feature_downloads_release(AbstractFetchDownloadService.kt:19)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.startDownloadJob$feature_downloads_release(AbstractFetchDownloadService.kt:3)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService$onStartCommand$1.invokeSuspend(AbstractFetchDownloadService.kt:5)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:2)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:19)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:14)
```

### PR DESCRIPTION
**Pull request by Amejia481**
See for the full explanation of the issue #6797 

---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt [hunk 1]
```diff
@@ -369,13 +370,13 @@ abstract class AbstractFetchDownloadService : Service() {
 
         response.body.useStream { inStream ->
             val newDownloadState = download.withResponse(response.headers, inStream)
-            downloadJobs[download.id]?.state = newDownloadState
+            currentDownloadJobState.state = newDownloadState
 
             useFileStream(newDownloadState, isResumingDownload) { outStream ->
-                copyInChunks(downloadJobs[download.id]!!, inStream, outStream)
+                copyInChunks(currentDownloadJobState, inStream, outStream)
```
Comment (Amejia481) [inline components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:376]: I think this ‼️ could be the source of [this crash ](https://sentry.prod.mozaws.net/operations/fenix-nightly/issues/7253086/?query=is:unresolved%20AbstractFetchDownloadService). What I'm not completely sure is how we got it that state as don't see any place where we are removing thing from the 
```

kotlin.KotlinNullPointerException: null
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.useFileStreamScopedStorage(AbstractFetchDownloadService.kt:18)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.useFileStream$feature_downloads_release(AbstractFetchDownloadService.kt:8)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService$performDownload$1.invoke(AbstractFetchDownloadService.kt:20)
    at mozilla.components.concept.fetch.Response$Body.useStream(Response.kt:1)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.performDownload$feature_downloads_release(AbstractFetchDownloadService.kt:19)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService.startDownloadJob$feature_downloads_release(AbstractFetchDownloadService.kt:3)
    at mozilla.components.feature.downloads.AbstractFetchDownloadService$onStartCommand$1.invokeSuspend(AbstractFetchDownloadService.kt:5)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:2)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:19)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:14)

```

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt [hunk 2]
```diff
@@ -510,17 +511,18 @@ abstract class AbstractFetchDownloadService : Service() {
     @TargetApi(Build.VERSION_CODES.P)
     @Suppress("Deprecation")
     private fun useFileStreamLegacy(download: DownloadState, append: Boolean, block: (OutputStream) -> Unit) {
+        val fileName = download.fileName ?: throw IllegalStateException("A fileName for a download is required")
```
Comment (kglazko) [inline components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:514]: Should we be throwing an exception currentDownloadJobState as well insetad of the return?

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt (hunk 1 above)
Comment (kglazko) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:376]: I think we should add a comment explaining why we made these changes and what we aren't sure of so people working on this in the future will have a trail to follow.

**Comment by kglazko** [comment]
Looks good to me- like everything we discussed in the pair programming session. I added a few nit suggestions. A note- we will need a separate PR to renable the UI test to see if this fixes the intermittent failures after this lands.

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt (hunk 1 above)
Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:376]: 👍

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt (hunk 2 above)
Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:514]: That's good question, if we do that will be will crashing the app but we will be discovering an illegal state in which the app shouldn't be, we should weight that 👍

Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:514]: We could keep as it is to collect more information and see it the crash it related, what do you think?

Comment (kglazko) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:514]: That sounds good to me! Let's keep as-is.

Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:514]: 👍

**Comment by Amejia481** [comment]
bors r=kglazko

**Pull request merged**